imports/server/MigrationRegistry.ts

```diff
@@ -0,0 +1,200 @@
+      return true;
+    } else if (observedVersion > target) {
+      this.log(`Migrations: not migrating, observed version ${observedVersion} exceeds target ${target}`);
+      return false;
+    }
+    // by elimination, observedVersion < target.  We have migrations to run.
+    // Acquire the lock.  We might race with another backend that is also
+    // trying to apply this same migration.
+    if (!this.lock()) {
+      this.log('Not migrating, control is locked');
+      return false;
+    } else {
+      // We hold the lock.  Let's do some work.
+      // Load the version again, since the first time we read it we weren't holding the lock.
+      let applied = this.getVersion();
+      while (applied < target) {
+        // Remember, this.migrations is 0-indexed, while the migration
+        // versions are 1-indexed so the DB contains the number of migrations
+        // completed
+        const nextMigration = this.migrations[applied];
+        this.log(`Migrations: running migration ${nextMigration.version} "${nextMigration.name}"`);
+        // Run the next migration in sequence.
+        nextMigration.up();
+        // Save the fact that we ran the migration durably.
+        const modified = this.collection.update({
+          _id: 'control',
+          locked: true,
+          version: applied,
+        }, {
+          $set: {
+            version: nextMigration.version,
+          },
+        });
+        if (modified !== 1) {
+          throw new Error(`Couldn't record completion of migration ${nextMigration.version}`);
+        }
+        applied += 1;
+      }
+
+      // We've finished applying migrations.  Release the lock and report success.
+      this.unlock();
+      return true;
+    }
+  }
+
+  getVersion(): number {
+    // The value in the database represents the last migration that has
+    // completed successfully, which (as a result of the structure of migration
+    // numbering) is equal to the number of migrations that have run.
+    //
+    // If there is no record in the database, no migrations have run.  We will
+    // backfill a record to give us a place to store a lock and update the
+    // version.
+    //
+    // If there is a record in the database, then its version indicates the
+    // number of migrations that have been run to completion successfully.
+    const control = this.collection.findOne({ _id: 'control' });
+    if (control === undefined) {
+      this.ensureControlCreated();
+      // Query again because we're not holding a lock.
+      return this.getVersion();
+    } else {
+      return control.version;
+    }
+  }
+
+  private ensureControlCreated() {
+    this.log('Migrations: creating control record at version 0');
+    // We use insert rather than upsert here
+    ignoringDuplicateKeyErrors(() => {
+      this.collection.insert({
+        _id: 'control',
+        version: 0,
+        locked: false,
+      });
+    });
```

Comment:
Huh, I'm surprised that this works. I thought that conflicts on `_id` threw a different error code than conflicts on secondary indexes, and I didn't think that I wrote `ignoringDuplicateKeyErrors` to catch both. Maybe Mongo finally fixed that.

Reply:
It's entirely possible that it doesn't!  I should test this.